src/python/arcor2_arserver/lock.py

```diff
@@ -89,103 +90,143 @@ def write_lock(self, obj_id: str, owner: str, lock_tree: bool) -> bool:
         return self._scene
 
     @scene.setter
-    def scene(self, scene: Optional["UpdateableCachedScene"] = None):
+    def scene(self, scene: Optional[UpdateableCachedScene] = None) -> None:
 
         assert self.project is None
 
         self._scene = scene
 
     @property
-    def project(self):
+    def project(self) -> Optional[UpdateableCachedProject]:
         return self._project
 
     @project.setter
-    def project(self, project: Optional["UpdateableCachedProject"] = None):
+    def project(self, project: Optional[UpdateableCachedProject] = None) -> None:
 
         assert self.scene is None
 
         self._project = project
 
-    async def read_lock(self, obj_id: str, owner: str) -> bool:
+    async def read_lock(self, obj_ids: Union[List[str], str], owner: str) -> bool:
 
-        root_id = self.get_root_id(obj_id)
+        if isinstance(obj_ids, str):
+            obj_ids = [obj_ids]
+
+        roots = [self.get_root_id(obj_id) for obj_id in obj_ids]
+        locked = []
 
         async with self._lock:
-            lock_record = self._get_lock_record(root_id)
-            locked = lock_record.read_lock(obj_id, owner)
+            for i, obj_id in enumerate(obj_ids):
+                lock_record = self._get_lock_record(roots[i])
+                acquired = lock_record.read_lock(obj_id, owner)
 
-            if not locked and lock_record.is_empty():
-                del self._locked_objects[root_id]
+                if acquired:
+                    locked.append(obj_id)
+                else:
+                    self._read_unlock(roots, locked, owner)
+                    return False
+        return True
 
-        return locked
+    async def read_unlock(self, obj_ids: Union[List[str], str], owner: str) -> None:
 
-    async def read_unlock(self, obj_id: str, owner: str) -> bool:
+        if isinstance(obj_ids, str):
+            obj_ids = [obj_ids]
 
-        root_id = self.get_root_id(obj_id)
+        roots = [self.get_root_id(obj_id) for obj_id in obj_ids]
 
         async with self._lock:
-            lock_record = self._get_lock_record(root_id)
-            unlocked = lock_record.read_unlock(obj_id, owner)
+            self._read_unlock(roots, obj_ids, owner)
 
-            if unlocked and lock_record.is_empty():
-                del self._locked_objects[root_id]
+    def _read_unlock(self, roots: List[str], obj_ids: List[str], owner: str) -> None:
+        """Internal function when lock is acquired"""
 
-        return unlocked
+        assert self._lock.locked()
 
-    async def write_lock(self, obj_id: str, owner: str, lock_tree: bool = False) -> bool:
+        for i, obj_id in enumerate(obj_ids):
+            try:
+                lock_record = self._get_lock_record(roots[i])
+                lock_record.read_unlock(obj_id, owner)
 
-        root_id = self.get_root_id(obj_id)
+                if lock_record.is_empty():
+                    del self._locked_objects[roots[i]]
+            except Exception:
```

Comment:
Tady by určitě mělo být `LockingException`. Takto kdybyste v kódu udělal chybu, tak vám to "sežere" i ValueError, IndexError, dělení nulou... Cokoliv. Bug v aplikaci by se tímto zamaskoval.

Reply:
Tady to je právě záměrně, když nastane výjimka tady, tak je uričtě něco hodně špatně, takže by aplikace neměla vůbec pokračovat. Kód v odchycení výjimky se pokouší jen o uvolnění zbytku zámků a výjimku opět raisne, takže nic nesežere...

Reply:
Ok, toho následujícího raisnutí jsem si nevšiml. Je to jak píšete, aplikace by měla skončit. A proto bych skoro řekl, že nemá cenu nic uvolňovat. Jediné správné řešení je padnout :)